clearbit/clearbit.go

```diff
@@ -18,13 +19,43 @@ type Client struct {
 	Reveal       *RevealService
 }
 
+// Config represents all the parameters available to configure a Clearbit
+// client
+type Config struct {
+	apiKey     string
+	httpClient *http.Client
+}
+
+// SetHTTPClient sets the optional http.Client we can use to make requests
+func SetHTTPClient(httpClient *http.Client) func(*Config) {
+	return func(config *Config) {
+		config.httpClient = httpClient
+	}
+}
+
+// SetAPIKey sets the Clearbit API key.
+//
+// When this is not provided we'll default to the `CLEARBIT_KEY` environment
+// variable.
+func SetAPIKey(apiKey string) func(*Config) {
+	return func(config *Config) {
+		config.apiKey = apiKey
+	}
+}
+
 // NewClient returns a new Client.
-func NewClient(httpClient *http.Client, apiKey string) *Client {
-	base := sling.New().Client(httpClient)
-	base.SetBasicAuth(apiKey, "")
+func NewClient(options ...func(*Config)) *Client {
+	config := Config{apiKey: os.Getenv("CLEARBIT_KEY")}
+
+	for _, option := range options {
```

Comment:
nice find. I haven't seen this style of configuration before. 

For future readers:
https://commandcenter.blogspot.com.uy/2014/01/self-referential-functions-and-design.html
https://dave.cheney.net/2016/11/13/do-not-fear-first-class-functions